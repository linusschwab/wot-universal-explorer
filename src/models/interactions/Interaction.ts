import {Link} from "../links";
import {OutputData} from "../data";

export class Interaction {

    public name: string;
    public base: string;
    public links: Link[];
    public outputData: OutputData;

    constructor(name: string) {
        this.name = name;
        this.links = [];
    }

    public registerLink(link: Link) {
        this.links.push(link);
        link.interaction = this;
    }

    public toString() {
        return this.name + ' interaction';
    }

}
