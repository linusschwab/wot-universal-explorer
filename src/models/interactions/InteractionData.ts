export class InteractionData {

    public timestamp: number;
    public data: any;

    constructor(data: any) {
        this.timestamp = Date.now();
        this.data = data;
    }

    public toString() {
        return {
            time: this.date,
            data: this.data
        }
    }

    public equals(data: InteractionData) {
        return JSON.stringify(this.data) === JSON.stringify(data.data);
    }

    get time() {
        return new Date(this.timestamp).toLocaleTimeString('de-CH');
    }

    get date() {
        return new Date(this.timestamp).toLocaleString('de-CH');
    }
}
