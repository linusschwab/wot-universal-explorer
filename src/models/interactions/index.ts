export {InteractionData} from './InteractionData';
export {InteractionPattern} from './InteractionPattern';
export {Action} from './Action';
export {Event} from './Event';
export {Property} from './Property';
export {ISubscriber} from './ISubscriber';
