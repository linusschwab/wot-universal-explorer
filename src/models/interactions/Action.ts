import {Interaction} from "./Interaction";
import {InputData, OutputData} from "../data";


export class Action extends Interaction {

    public inputData: InputData;

    constructor(name: string, outputData: OutputData, inputData: InputData) {
        super(name);
        this.outputData = outputData;
        this.inputData = inputData;
    }

    public invoke() {
        // TODO: Implement
        return true;
    }

    public toString() {
        return this.name + ' action';
    }

}
