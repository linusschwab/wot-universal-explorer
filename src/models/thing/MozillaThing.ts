import * as WebSocket from "ws";

import {Thing} from "./Thing";
import {WebSocketManager} from "../../controllers";
import {InteractionData} from "../interactions";


export class MozillaThing extends Thing {

    public ws: WebSocket;

    constructor(name: string, type: string, base: string) {
        super(name, type, base);
        this.setupWebSocket();
    }

    private setupWebSocket() {
        let wsUrl = this.base.replace('http://', 'ws://');
        if (this.base.includes(process.env.MOZ_BASE)) {
            wsUrl += '?jwt=' + process.env.MOZ_AUTH;
        }

        // TODO: Handle error(s)
        this.ws = new WebSocket(wsUrl);

        this.ws.on('message', data => this.wsHandleMessage(data));
    }

    public wsHandleMessage(data: WebSocket.Data) {
        let message: any;
        try {
            message = WebSocketManager.parseMessage(data);
        } catch (e) {
            return;
        }

        if (message.messageType === 'propertyStatus') {
            for (let propertyName in message.data) {
                try {
                    const property = this.getProperty(propertyName);
                    const data = new InteractionData(message.data[propertyName]);
                    property.update(data);
                } catch (e) {
                    // Do nothing
                }
            }
        }
    }
}
