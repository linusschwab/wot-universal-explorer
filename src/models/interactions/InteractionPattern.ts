import * as WebSocket from "ws";
import * as getSlug from "speakingurl";

import {Link, Operation} from "../links";
import {Thing} from "../thing";
import {InteractionData} from "./InteractionData";
import {InteractionError} from "../../tools/errors";


export abstract class InteractionPattern {

    public name: string;
    public description: string;

    public thing: Thing;
    public base: string;

    public links: Link[];
    public url: string;
    public operations: Operation[];

    public data: InteractionData[];
    public subscribers: WebSocket[];

    constructor(name: string) {
        this.name = name;
        this.links = [];
        this.data = [];
        this.subscribers = [];
    }

    public registerLink(link: Link) {
        this.links.push(link);
        link.interaction = this;
    }

    public subscribe(ws: WebSocket) {
        if (this.subscribers.includes(ws)) {
            throw new InteractionError('Already subscribed to interaction');
        }

        this.subscribers.push(ws);
    }

    public unsubscribe(ws: WebSocket) {
        if (!this.subscribers.includes(ws)) {
            throw new InteractionError('Not subscribed to interaction');
        }

        let index = this.subscribers.indexOf(ws);
        this.subscribers.splice(index, 1);
    }

    public abstract async notifySubscribers(data: InteractionData): Promise<void>;

    public async storeData(data: InteractionData) {
        // Delete old items if too many
        if (this.data.length > 1000) {
            this.data.splice(0, 1);
        }

        this.data.push(data);
    }

    public newData(data: InteractionData): boolean {
        return this.data.length == 0 || !data.equals(this.data[this.data.length-1])
    }

    public abstract toString(): string;

    get slug() {
        return getSlug(this.name);
    }
}
