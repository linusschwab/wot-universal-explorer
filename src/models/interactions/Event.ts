import {InteractionPattern} from "./InteractionPattern";
import {DataSchema} from "../schema";
import {Operation} from "../links";
import {InteractionData} from "./InteractionData";
import {TimeoutError} from "../../tools/errors";


export class Event extends InteractionPattern {

    public schema: DataSchema;

    constructor(name: string, schema: DataSchema) {
        super(name);
        this.schema = schema;
    }

    public async poll() {
        // TODO: Choose correct link
        return this.links[0].execute(null);
    }

    public async update() {
        // Only poll if regular event link
        if (this.links.length === 0) {
            return;
        }

        try {
            const data = new InteractionData(await this.poll());

            // Check if data changed
            if (this.newData(data)) {
                this.storeData(data);
                this.notifySubscribers(data);
            }
        } catch (e) {
            if (e instanceof TimeoutError === false) {
                throw e;
            }
        }
    }

    public getData(newerThan: number = 0, limit: number = 0) {
        let newData: InteractionData[] = [];
        if (newerThan > 0) {
            for (let data of this.data) {
                if (data.timestamp > newerThan) {
                    newData.push(data);
                }
            }
        } else {
            newData = this.data;
        }
        return newData.slice(-limit);
    }

    public async notifySubscribers(data: InteractionData) {
        for (let ws of this.subscribers) {
            ws.send(JSON.stringify({
                messageType: 'event',
                data: data.toString()
            }));
        }
    };

    public toString() {
        return 'Event ' + this.name;
    }

    public get url() {
        return '/' + this.thing.slug + '/events/' + this.slug;
    }

    public get operations(): Operation[] {
        return [new Operation('get', 'List of ' + this.name + ' Event data', this)];
    }
}
