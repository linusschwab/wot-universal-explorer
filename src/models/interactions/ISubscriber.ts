import {InteractionData} from "./InteractionData";
import {InteractionPattern} from "./InteractionPattern";

export interface ISubscriber {
    (interaction: InteractionPattern, data: InteractionData) : void;
}
