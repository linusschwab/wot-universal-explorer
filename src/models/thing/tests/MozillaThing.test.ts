import * as WebSocket from "ws";

import {MozillaThing} from "../MozillaThing";
import {Property} from "../../interactions";


// Create mocks
jest.mock('ws');
let mockWs: WebSocket;

beforeAll(() => {
    // Mock implementations
    const MockWebSocket = jest.fn<WebSocket>(() => ({
        send: jest.fn()
    }));
    mockWs = new MockWebSocket();
});

test('property status update sends websocket notification', async () => {
    const thing = new MozillaThing('TestThing', 'Thing', 'http://localhost/');
    const property = new Property('on', null, false, true);
    thing.registerInteraction(property);
    thing.subscribeToProperty('on', mockWs);

    const data = '{"messageType": "propertyStatus", "data": {"on": false}}';
    await thing.wsHandleMessage(data);

    expect(mockWs.send).toMatchSnapshot();
});
