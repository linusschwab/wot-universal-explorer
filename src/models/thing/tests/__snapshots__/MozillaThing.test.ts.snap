// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`property status update sends websocket notification 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"messageType\\":\\"propertyStatus\\",\\"data\\":{\\"on\\":false}}",
    ],
  ],
}
`;
